File: agent-gui/lib/guide-markdown-section-parser.ts
/** Splits guide markdown into indexable sections (aligns with AgentModel GuideMarkdownSectionParser). */

export type MarkdownSection = {
  heading: string;
  slug: string;
  body: string;
};

const H2_RE = /^##\s+(.+)$/gm;
const H1_RE = /^#\s+(.+)$/gm;

function slugify(heading: string): string {
  const slug = heading
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff]+/g, "-")
    .replace(/^-+|-+$/g, "");
  if (!slug) return "section";
  return slug.length > 80 ? slug.slice(0, 80).replace(/-+$/, "") : slug;
}

/** ## sections for workflow-style guides. */
export function parseH2Sections(markdown: string): MarkdownSection[] {
  if (!markdown?.trim()) return [];

  const matches = [...markdown.matchAll(H2_RE)];
  if (matches.length === 0) return [];

  return matches.map((match, i) => {
    const heading = match[1]?.trim() ?? "";
    const bodyStart = match.index! + match[0].length;
    const bodyEnd =
      i + 1 < matches.length ? matches[i + 1]!.index! : markdown.length;
    return {
      heading,
      slug: slugify(heading),
      body: markdown.slice(bodyStart, bodyEnd).trim(),
    };
  });
}

/**
 * # sections after document title — KC module docs use H1 for major sections.
 * Falls back to ## when only one top-level # exists.
 */
export function parseIndexableSections(markdown: string): MarkdownSection[] {
  if (!markdown?.trim()) return [];

  const h1Matches = [...markdown.matchAll(H1_RE)];
  if (h1Matches.length > 1) {
    return h1Matches.slice(1).map((match, i) => {
      const heading = match[1]?.trim() ?? "";
      const bodyStart = match.index! + match[0].length;
      const next = h1Matches[i + 2];
      const bodyEnd = next?.index ?? markdown.length;
      return {
        heading,
        slug: slugify(heading),
        body: markdown.slice(bodyStart, bodyEnd).trim(),
      };
    });
  }

  const h2 = parseH2Sections(markdown);
  if (h2.length > 0) return h2;

  return [];
}

export function tryExtractSectionBody(
  markdown: string,
  sectionHeading: string,
): string | null {
  if (!sectionHeading.trim()) return null;
  const target = sectionHeading.trim().toLowerCase();
  for (const section of parseIndexableSections(markdown)) {
    if (section.heading.trim().toLowerCase() === target) {
      return section.body;
    }
  }
  return null;
}

File: QuickerRpc/src/QuickerRpc.AgentModel/Search/GuideMarkdownSectionParser.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace QuickerRpc.AgentModel.Search;

/// <summary>Splits guide markdown into ## sections for fragment indexing.</summary>
public static class GuideMarkdownSectionParser
{
    private static readonly Regex H2Regex = new(
        @"^##\s+(.+)$",
        RegexOptions.Multiline | RegexOptions.CultureInvariant);

    public sealed class Section
    {
        public string Heading { get; init; } = string.Empty;

        public string Slug { get; init; } = string.Empty;

        public string Body { get; init; } = string.Empty;
    }

    public static IReadOnlyList<Section> ParseH2Sections(string markdown)
    {
        if (string.IsNullOrWhiteSpace(markdown))
        {
            return Array.Empty<Section>();
        }

        var matches = H2Regex.Matches(markdown);
        if (matches.Count == 0)
        {
            return Array.Empty<Section>();
        }

        var sections = new List<Section>(matches.Count);
        for (var i = 0; i < matches.Count; i++)
        {
            var heading = matches[i].Groups[1].Value.Trim();
            var bodyStart = matches[i].Index + matches[i].Length;
            var bodyEnd = i + 1 < matches.Count ? matches[i + 1].Index : markdown.Length;
            var body = markdown.Substring(bodyStart, bodyEnd - bodyStart).Trim();
            sections.Add(new Section
            {
                Heading = heading,
                Slug = Slugify(heading),
                Body = body,
            });
        }

        return sections;
    }

    public static string? TryExtractSectionBody(string markdown, string sectionHeading)
    {
        if (string.IsNullOrWhiteSpace(markdown) || string.IsNullOrWhiteSpace(sectionHeading))
        {
            return null;
        }

        foreach (var section in ParseH2Sections(markdown))
        {
            if (string.Equals(section.Heading, sectionHeading, StringComparison.OrdinalIgnoreCase))
            {
                return section.Body;
            }
        }

        return null;
    }

    internal static string Slugify(string heading)
    {
        var lower = heading.ToLowerInvariant();
        var slug = Regex.Replace(lower, @"[^a-z0-9\u4e00-\u9fff]+", "-").Trim('-');
        if (slug.Length > 80)
        {
            slug = slug.Substring(0, 80).TrimEnd('-');
        }

        return string.IsNullOrEmpty(slug) ? "section" : slug;
    }
}
